### ISSUE TITLE
Gradient for d/db a**b should be 0 at a=0 instead of NaN

### ISSUE THREAD

**Issue by lukas-schott** [pytorch/pytorch#27134]
## 🐛 Bug

The derivative `d/db a**b` at `a = 0` should be 0. This can be seen by doing `d/db a**b = a**b  log(a)` and looking at `lim a-->  0` (works in tf).

## To Reproduce

```
a = torch.tensor([0., 2.])
b = torch.ones((2), requires_grad=True)
c = torch.sum(a**b)
c.backward()
print('grad b', b.grad)    # is [nan, 1.386] should be [0, 1.386]
```

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```

PyTorch version: 1.2.0.dev20190722
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration: GPU 0: Tesla P100-PCIE-16GB
Nvidia driver version: 410.79
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.2

Versions of relevant libraries:
[pip3] advertorch==0.1.5
[pip3] numpy==1.16.1
[pip3] torch==1.2.0
[pip3] torch-nightly==1.2.0.dev20190722
[pip3] torchvision==0.4.0


cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by soumith** [comment]
cc: @albanD

**Comment by albanD** [comment]
This is only true if `b > 0` right?
What would be the expected value when `b = 0`?

**Comment by lukas-schott** [comment]
> This is only true if `b > 0` right?

yes, I guess so. 


> What would be the expected value when `b = 0`?

So `d/db a**b` for `a=0` and `b=0`, the derivative should be `nan`.

This can be seen by approach limits from `a`, and `b` which lead `-inf` and `0`. However, I'm not sure because tf has `0` for this scenario though.)

**Comment by albanD** [comment]
I would be happy with the following:
at `a = 0`:
- if b < 0: the gradient is `+/- inf * -inf = nan`.
- if b = 0: the gradient is `1 * -inf = -inf`. (not sure why tf has 0 here either)
- if b > 0: the gradient is `0`.

I would be happy to review a PR that adds this :)

**Comment by agolynski** [comment]
I am taking on this issue. Discussed offline with Alban re: edge cases.

We decided that it might be better to define d(a^b)/db at a = 0 and b < 0 to be -inf. This is due to d(a^b)/db -> -inf for a fixed b as a -> +0
Currently, tensorflow defines d(a^b)/db = nan for a = 0 and b < 0, so we are in a slight disagreement here.

We can define d(a^b)/db = 0 for a = 0 and b = 0 by continuity as a^b = 0 along a = 0 for b>0 and derivative db(a^b)/db tends to 0 as we approach from a -> +0 for any b>0.
Currently, tensorflow agrees with us.

Landed PRs:
https://github.com/pytorch/pytorch/pull/32062
https://github.com/pytorch/pytorch/pull/32063

**Comment by agolynski** [comment]
Closing issue for now. Let me know if you need further changes.
